### ISSUE TITLE
Building with cabal alone

### ISSUE THREAD

**Issue by jrp2014** [lehins/massiv#71]
To build with `cabal alone` I just added a `cabal.project` file at the top level with the following content
```
packages:   */*.cabal
optimization: 2

package    massiv-bench
  benchmarks: True
```
at the top level, and then you can `cabal v2-build all`, `cabal v2-bench massiv-bench`, `cabal v2-test massiv`, etc.

The examples seem to need at least one extra external library
```
sudo apt install freeglut3 freeglut3-dev
```

The examples still don't build completely:
```
[12 of 16] Compiling Data.Massiv.Array.IO.Base ( src/Data/Massiv/Array/IO/Base.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/massiv-io-0.1.5.0/opt/build/Data/Massiv/Array/IO/Base.o )
[13 of 16] Compiling Data.Massiv.Array.IO.Image.JuicyPixels ( src/Data/Massiv/Array/IO/Image/JuicyPixels.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/massiv-io-0.1.5.0/opt/build/Data/Massiv/Array/IO/Image/JuicyPixels.o )

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:775:3: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In a case alternative: Patterns not matched: (JP.ImageY32 _)
    |
775 |   case jpDynImg of
    |   ^^^^^^^^^^^^^^^^...

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:821:3: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In a case alternative: Patterns not matched: (JP.ImageY32 _)
    |
821 |   case jpDynImg of
    |   ^^^^^^^^^^^^^^^^...

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:852:17: warning: [-Wdeprecations]
    In the use of ‘computeSource’
    (imported from Data.Massiv.Array, but defined in massiv-0.3.0.0:Data.Massiv.Array.Manifest.Internal):
    Deprecated: "In favor of less restrictive `convert`"
    |
852 |     [ (\Refl -> computeSource img) <$>
    |                 ^^^^^^^^^^^^^

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:881:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘showJP’: Patterns not matched: (JP.ImageY32 _)
    |
881 | showJP (JP.ImageY8     _) = "Image S Y Word8"
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:906:11: warning: [-Wdeprecations]
    In the use of ‘computeSource’
    (imported from Data.Massiv.Array, but defined in massiv-0.3.0.0:Data.Massiv.Array.Manifest.Internal):
    Deprecated: "In favor of less restrictive `convert`"
    |
906 |   !arrS = computeSource img :: Image S cs (JP.PixelBaseComponent a)
    |           ^^^^^^^^^^^^^

src/Data/Massiv/Array/IO/Image/JuicyPixels.hs:972:12: warning: [-Wdeprecations]
    In the use of ‘fromVector’
    (imported from Data.Massiv.Array.Manifest.Vector):
    Deprecated: "In favor of safer `fromVectorM`"
    |
972 |   return $ fromVector Par (Sz (m :. n)) $ V.unsafeCast v
    |            ^^^^^^^^^^
[14 of 16] Compiling Data.Massiv.Array.IO.Image.Netpbm ( src/Data/Massiv/Array/IO/Image/Netpbm.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/massiv-io-0.1.5.0/opt/build/Data/Massiv/Array/IO/Image/Netpbm.o )

src/Data/Massiv/Array/IO/Image/Netpbm.hs:169:12: warning: [-Wdeprecations]
    In the use of ‘fromVector’
    (imported from Data.Massiv.Array.Manifest.Vector):
    Deprecated: "In favor of safer `fromVectorM`"
    |
169 |   return $ fromVector Par (Sz (m :. n)) $ V.unsafeCast v
    |            ^^^^^^^^^^
[15 of 16] Compiling Data.Massiv.Array.IO.Image ( src/Data/Massiv/Array/IO/Image.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/massiv-io-0.1.5.0/opt/build/Data/Massiv/Array/IO/Image.o )
[16 of 16] Compiling Data.Massiv.Array.IO ( src/Data/Massiv/Array/IO.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/massiv-io-0.1.5.0/opt/build/Data/Massiv/Array/IO.o )
Configuring library for examples-0.1.0.0..
Preprocessing library for examples-0.1.0.0..
Building library for examples-0.1.0.0..
[1 of 2] Compiling Examples.Convolution ( src/Examples/Convolution.hs, /home/jrp/Projects/massiv/dist-newstyle/build/x86_64-linux/ghc-8.6.4/examples-0.1.0.0/opt/build/Examples/Convolution.o )

src/Examples/Convolution.hs:14:41: error:
    • Couldn't match expected type ‘Sz Ix2’ with actual type ‘Ix2’
    • In the second argument of ‘makeArray’, namely ‘arrSz’
      In the expression: makeArray comp arrSz lightFunc
      In an equation for ‘arrLightIx2’:
          arrLightIx2 comp arrSz
            = makeArray comp arrSz lightFunc
            where
                lightFunc (i :. j)
                  = sin (fromIntegral (i ^ (2 :: Int) + j ^ (2 :: Int)) :: Double)
   |
14 | arrLightIx2 comp arrSz = makeArray comp arrSz lightFunc
   |                                         ^^^^^

src/Examples/Convolution.hs:20:33: error:
    • Couldn't match expected type ‘Sz Ix2’ with actual type ‘Ix2’
    • In the first argument of ‘makeStencil’, namely ‘(3 :. 3)’
      In the expression: makeStencil (3 :. 3) (1 :. 1)
      In the expression:
        makeStencil (3 :. 3) (1 :. 1)
          $ \ get
              -> (get (- 1 :. - 1) + get (- 1 :. 0) + get (- 1 :. 1)
                    + get (0 :. - 1)
                    + get (0 :. 0)
                    + get (0 :. 1)
                    + get (1 :. - 1)
                    + get (1 :. 0)
                    + get (1 :. 1))
                   / 9
   |
20 | average3x3Filter = makeStencil (3 :. 3) (1 :. 1) $ \ get ->
   |                                 ^^^^^^

src/Examples/Convolution.hs:28:40: error:
    • Couldn't match expected type ‘Sz Ix2’ with actual type ‘Ix2’
    • In the first argument of ‘makeConvolutionStencil’, namely
        ‘(3 :. 3)’
      In the expression: makeConvolutionStencil (3 :. 3) (1 :. 1)
      In the expression:
        makeConvolutionStencil (3 :. 3) (1 :. 1)
          $ \ get
              -> get (- 1 :. - 1) 1
                   . get (- 1 :. 0) 1
                       . get (- 1 :. 1) 1
                           . get (0 :. - 1) 1
                               . get (0 :. 0) 1
                                   . get (0 :. 1) 1
                                       . get (1 :. - 1) 1 . get (1 :. 0) 1 . get (1 :. 1) 1
   |
28 | sum3x3Filter = makeConvolutionStencil (3 :. 3) (1 :. 1) $ \ get ->
   |                                        ^^^^^^
Installing   OpenGLRaw-3.3.2.0 (lib)
Completed    OpenGLRaw-3.3.2.0 (lib)
cabal: Failed to build examples-0.1.0.0 (which is required by exe:mkfilter
from examples-0.1.0.0, exe:life from examples-0.1.0.0 and others).
```
These are fixable by adding a few `Sz` annotations.  The examples also seem to need `massiv-scheduler` as a dependency.

**Comment by lehins** [comment]
That is all correct.
I am getting very close to releasing new version `v0.3.0`, which all these build failures are related to.
Thank you for pointing them out, I did have it in my plans to update examples and readme, but now it will be recorded as a ticket. :) `massiv-scheduler` is also not yet released to hackage, but is inside this repo, I don't know what needs to be done in order for cabal to pick it up though.

**Comment by lehins** [comment]
So, @jrp2014 the `cabal.project` has been added in your #73 I fixed all of the examples in 01149e8a6d43657c5b8c14851b10383a9aff6711 and also added dependency on freeglut instructions to the examples README.md
I think we can consider this ticket taken care of, if I missed something feel free to reopen.
